examples/ShowPlayer.cpp

```diff
@@ -66,90 +69,184 @@ int ShowPlayer::Init() {
+  PlaybackTime playhead_time = m_playback_pos;
+  ShowLoader::State state;
+  bool found = false;
+  while (true) {ShowEntry entry;
+    state = m_loader.NextEntry(&entry);
+    switch (state) {
+      case ShowLoader::END_OF_FILE:
+        OLA_FATAL << "Show file ends before the start time (Actual length "
+                  << m_playback_pos << " ms)";
+        return state;
+      case ShowLoader::INVALID_LINE:
+        HandleInvalidLine();
+        return state;
+      default: {
+      }
+    }
+    playhead_time += entry.next_wait;
+    entries[entry.universe] = entry;
+    if (!found && playhead_time == seek_time) {
+      // Use the next frame if landing on the trailing edge of a frame's timeout
+      found = true;
+      continue;
+    }
+    if (found || playhead_time > seek_time) {
+      break;
+    }
+  }
+  m_playback_pos = seek_time;
+
+  // Send data in the state it would be in at the given time
+  unordered_map<unsigned int, ShowEntry>::iterator entry_it;
+  for (entry_it = entries.begin(); entry_it != entries.end(); ++entry_it) {
+    SendFrame(entry_it->second);
   }
+  // Adjust the timeout to handle landing in the middle of the entry's timeout
+  RegisterNextTimeout(playhead_time-seek_time);
+
+  return ShowLoader::OK;
 }
 
 
 /**
- * Get the next time offset
+ * Send the next frame in the show file
  */
-ShowLoader::State ShowPlayer::RegisterNextTimeout() {
-  unsigned int timeout;
-  ShowLoader::State state = m_loader.NextTimeout(&timeout);
-  if (state != ShowLoader::OK) {
-    return state;
+void ShowPlayer::SendNextFrame() {
+  ShowEntry entry;
+  ShowLoader::State state = m_loader.NextEntry(&entry);
+  if (state == ShowLoader::END_OF_FILE ||
+      (m_stop > 0 && m_playback_pos >= m_stop)) {
+    if (entry.buffer.Size() > 0) {
+      // Send the last frame before looping/exiting
+      SendFrame(entry);
+    }
+    HandleEndOfFile();
+    return;
+  } else if (state == ShowLoader::INVALID_LINE) {
+    HandleInvalidLine();
+    return;
   }
+  SendEntry(entry);
+}
 
+
+/**
+ * Send @p entry, update playhead position, and wait for next
+ * @param entry the show file entry to send
+ */
+void ShowPlayer::SendEntry(const ShowEntry &entry) {
+  // Send DMX data
+  SendFrame(entry);
+
+  // Set when next to send data
+  m_playback_pos += entry.next_wait;
+  RegisterNextTimeout(entry.next_wait);
+}
+
+
+/**
+ * Send the next frame in @p timeout milliseconds
+ */
+void ShowPlayer::RegisterNextTimeout(const unsigned int timeout) {
   OLA_INFO << "Registering timeout for " << timeout << "ms";
   m_client.GetSelectServer()->RegisterSingleTimeout(
-      timeout,
-      ola::NewSingleCallback(this, &ShowPlayer::SendNextFrame));
-  return state;
+     timeout,
+     ola::NewSingleCallback(this, &ShowPlayer::SendNextFrame));
+}
+
+
+/**
+ * Send data contained in @p entry
+ */
+void ShowPlayer::SendFrame(const ShowEntry &entry) const {
+  OLA_INFO << "Universe: " << entry.universe << ": " << entry.buffer.ToString();
```

Comment:
This probably wants to be debug level to avoid flooding the log with every frame.

Reply:
It was info level before this PR - If this changes it may also make sense to change the log level for the timeouts.  Thoughts?

Reply:
Agreed that the timeouts should match the same log level.

Well you've been using it, how do you feel, is it too noisy or about right? Looking at the current live source, there aren't as many logging bits specifically in the ola_recorder stuff as I thought there might be, so it might not be an issue. Certainly there seem to be very few other bits at OLA_INFO level.

Reply:
I asked a non-programmer friend about this because sometimes it's nice to hear from people who don't care about the low-level details.  He thought knowing actual output was less important ("Want to know what the output is?  Look at the lights! Or open the web browser and see what olad says.").

I'm inclined to drop the DMX output and timeouts down to debug and open with a summary of how the file will be played (e.g. Playing from ##ms to ##ms, loop delay, etc) so the user knows it's alive.

The other things that log at info level should probably stay there, i.e. the looping notice.

Reply:
For the summary, what about running verify at the start, which also stops playback getting half way through and then failing. Although perhaps add a --no-verify option to playback for cases where you're e.g. running schedules on demand from an external system and don't want the delay (i.e. assuming you've pre-verified all your shows).

I was chatting to another user and they said similar, things like loop status at the start end end (I don't think we do a countdown of iterations currently) are what they care about. Which also allows for external application wrapping.

We could also use cout to write to stdout, although my feeling is that's probably not necessarily the right thing to do given this is almost a daemon, you don't interact with it or expect output from it, so the strength of being able to redirect all logs to syslog is probably higher.

Reply:
I like the idea of re-using verify like that.
I hadn't considered daemonizing the program like that, although it sounds nice - if run as a cron job or similar to playback at a certain time or something.

Reply:
> We could also use cout to write to stdout, although my feeling is that's probably not necessarily the right thing to do given this is almost a daemon, you don't interact with it or expect output from it, so the strength of being able to redirect all logs to syslog is probably higher.

My instinct is to output `--verify` to stdout (it's intentional and likely run interactively) and the pre-playback summary to log level info (as the rest of playback output already ends up in a logging target).  If we go this route, I'd add a louder note to the man page/help text about the `-l` options as otherwise it's entirely possible for a user to see no output at all, including summary, loop notices, etc. while running the program because the default log level is warn.

The only other place the logging level options are mentioned is in the docs for olad; because this seems to be the example program that makes heaviest use of the logging system it feels worth a mention.